### PR TITLE
Teach sn.exe about PE32+, and while there ROM images and MS-DOS-less images.

### PR DESCRIPTION
**Pull request by jaykrell**
Fixes https://github.com/mono/mono/issues/8218

### CODE REVIEW

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 1]
```diff
@@ -285,116 +285,233 @@ private UInt32 RVAtoPosition (UInt32 r, int sections, byte[] headers)
...
 		{
-			FileStream fs = File.OpenRead (fileName);
-			StrongNameSignature sn = StrongHash (fs, StrongNameOptions.Metadata);
-			fs.Close ();
-
-			return sn.Hash;
+			using (FileStream fs = File.OpenRead (fileName))
+			{
+				StrongNameSignature sn = StrongHash (fs, StrongNameOptions.Metadata);
+				fs.Close ();
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: I wondered about that. ok.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 2]
```diff
@@ -285,116 +285,233 @@ private UInt32 RVAtoPosition (UInt32 r, int sections, byte[] headers)
 			return 0;
 		}
 
-		internal StrongNameSignature StrongHash (Stream stream, StrongNameOptions options)
+		private StrongNameSignature Error (string a)
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: ok.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 3]
```diff
@@ -285,116 +285,233 @@ private UInt32 RVAtoPosition (UInt32 r, int sections, byte[] headers)
...
+				peHeader = BitConverterLE.ToInt32 (mz, 60);
+				if (peHeader < mzSize)
+					return Error ("peHeader_lt_64");
+
+				// Read MS-DOS stub.
+
+				if (peHeader > mzSize) {
+					int additionalMzSize = peHeader - mzSize;
+					byte[] mz2 = new byte [peHeader];
+					if (stream.Read (mz2, mzSize, additionalMzSize) != additionalMzSize)
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: An arrays be resized, and retain their contents? I didn't think so.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 4]
```diff
@@ -285,116 +285,233 @@ private UInt32 RVAtoPosition (UInt32 r, int sections, byte[] headers)
...
+							int after = (int)(length - s);
+							if (after > 0) {
+								cs.Write (section, s, after);
+							}
+						}
+						else
+							cs.Write (section, 0, length);
+					}
+
+					cs.Close ();
```
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: I suspected so, ok.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (marek-safar) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: Yes, that's how it works in .net

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: This is not redundant it turns out. At least not all of them, quite. You have to dispose the stream before getting the hash.

**Comment by jaykrell** [comment]
Ping?

**Comment by jaykrell** [comment]
@marek-safar good now?

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 5]
```diff
@@ -285,116 +285,230 @@ private UInt32 RVAtoPosition (UInt32 r, int sections, byte[] headers)
...
 		// return the same result as the undocumented and unmanaged GetHashFromAssemblyFile
 		public byte[] Hash (string fileName) 
 		{
-			FileStream fs = File.OpenRead (fileName);
-			StrongNameSignature sn = StrongHash (fs, StrongNameOptions.Metadata);
-			fs.Close ();
-
-			return sn.Hash;
+			using (FileStream fs = File.OpenRead (fileName))
+			{
```
Comment (filipnavara) [inline mcs/class/Mono.Security/Mono.Security/StrongName.cs]: Nitpick: Brace on the previous line.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 6]
```diff
@@ -433,7 +545,6 @@ public bool Verify (string fileName)
 			bool result = false;
 			using (FileStream fs = File.OpenRead (fileName)) {
 				result = Verify (fs);
-				fs.Close ();
```
Comment (filipnavara) [inline mcs/class/Mono.Security/Mono.Security/StrongName.cs]: You can remove the `result` variable and just `return Verify (fs);`.

File: mcs/class/Mono.Security/Mono.Security/StrongName.cs [hunk 7]
```diff
@@ -422,7 +535,6 @@ public bool Sign (string fileName)
 			using (FileStream fs = File.OpenWrite (fileName)) {
 				fs.Position = sn.SignaturePosition;
 				fs.Write (signature, 0, signature.Length);
-				fs.Close ();
 				result = true;
```
Comment (filipnavara) [inline mcs/class/Mono.Security/Mono.Security/StrongName.cs]: Just `return true` and get rid of the `result` variable?

Comment (jaykrell) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: I was wondering about that. In msvc C at least, you don't want to return from a __try, better to fall out. Return within __try runs exception handling code. I wasn't sure if that thinking was being applied here.

Comment (filipnavara) [reply mcs/class/Mono.Security/Mono.Security/StrongName.cs]: Just for the record, the compiler moves the return out. Consider the following C# code:

```csharp
int Test ()
{
	using (System.IO.File.OpenRead("aaa.text"))
		return 0;
}
```

translates into the following IL:

```
.method private hidebysig 
	instance int32 Test () cil managed 
{
	// Method begins at RVA 0x2050
	// Code size 29 (0x1d)
	.maxstack 1
	.locals init (
		[0] class [mscorlib]System.IO.FileStream,
		[1] int32
	)

	IL_0000: nop
	IL_0001: ldstr "aaa.text"
	IL_0006: call class [mscorlib]System.IO.FileStream [mscorlib]System.IO.File::OpenRead(string)
	IL_000b: stloc.0
	.try
	{
		IL_000c: ldc.i4.0
		IL_000d: stloc.1
		IL_000e: leave.s IL_001b
	} // end .try
	finally
	{
		IL_0010: ldloc.0
		IL_0011: brfalse.s IL_001a

		IL_0013: ldloc.0
		IL_0014: callvirt instance void [mscorlib]System.IDisposable::Dispose()
		IL_0019: nop

		IL_001a: endfinally
	} // end handler

	IL_001b: ldloc.1
	IL_001c: ret
} // end of method Program::Test
```

I'm not Jon Skeet to link to all the technical details in the specification, but I am sure that if the compiler considers it the same then it should be semantically equivalent. :-)

**Pull request merged**